### PR TITLE
[🛠️] Fix-whisper-breaking-changes 

### LINKED ISSUE #21878: Make whisper-event checkpoints compliant to support `return_timestamp`
### System Info

- `transformers` version: 4.27.0.dev0
- Platform: Linux-5.10.147+-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.12.1
- PyTorch version (GPU?): 1.13.1+cu116 (False)
- Tensorflow version (GPU?): 2.11.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: NO
- Using distributed or parallel set-up in script?: NO

### Who can help?

@sanchit-gandhi @ArthurZucker 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Inferring a Whisper checkpoint fine-tuned before the `TimestampstampProcessor` was introduced into transformers returns a rather un-informed error message `AttributeError: 'GenerationConfig' object has no attribute 'no_timestamps_token_id'`

Minimum steps to reproduce this:
```python
from transformers.pipelines import AutomaticSpeechRecognitionPipeline, pipeline
from datasets import load_dataset

cv11 = load_dataset("mozilla-foundation/common_voice_11_0", "hi", split="test", streaming=True)

pipe = pipeline(model="sanchit-gandhi/whisper-small-hi", return_timestamps=True)

test_sample = {"raw": next(iter(cv11))["audio"]["array"],
"sampling_rate": next(iter(cv11))["audio"]["sampling_rate"]}

pipe(test_sample)
```

Colab/ notebook: [here](https://github.com/Vaibhavs10/scratchpad/blob/main/pipeline_backward_compatability_test.ipynb)

The above snippet throws an error as mentioned above. This problem effects the majority (727) of the checkpoints fine-tuned during the Whisper Event.

P.S. This has been reported by multiple community members, so not just me.

### Expected behavior

We should ideally make the `return_timestamp` functionality backwards compatible or throw a more informative error message.

Sorry if there already is a way to do this and I am just misinformed.

### LINKED ISSUE #21925: [Whisper] Index error with model weights like base, large-v2
### System Info

- `transformers` version: 4.27.0.dev0
- Platform: Linux-5.4.0-66-generic-x86_64-with-glibc2.31
- Python version: 3.10.9
- Huggingface_hub version: 0.12.1
- PyTorch version (GPU?): 1.13.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@sanchit-gandhi @Narsil 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. I have been using Whisper from transformers library with my own custom script. The full code is attached below for reference:

```
import argparse
import base64
import os
import tempfile

import torch
from transformers import (
    AutoFeatureExtractor,
    AutoTokenizer,
    WhisperForConditionalGeneration,
    WhisperProcessor,
    pipeline,
)
from utils.b64 import convert_to_base64

model_dir = "openai/whisper-base"


class Whisper:
    def __init__(self):
        self.tokenizer = AutoTokenizer.from_pretrained(model_dir)
        self.feature_extractor = AutoFeatureExtractor.from_pretrained(model_dir)
        self.processor = WhisperProcessor.from_pretrained(model_dir)
        self.device = "cuda" if torch.cuda.is_available() else "cpu"
        # print([x for x in Path(model_dir).iterdir()])
        self.model = WhisperForConditionalGeneration.from_pretrained(model_dir).to(self.device)
        self.is_timestamp = False

    def predict_raw(self, payload):
            if payload is None:
                return {"inputerror": "JSON expected"}

            if "wav_base64" not in payload:
                return {"inputerror": "Missing key wav_base64 in payload."}


            afs = payload["wav_base64"]
            if not isinstance(afs, str):
                return {"inputerror": "Audio file to passed as input in base64 format"}

            if "timestamps" not in payload:
                timestamp = self.is_timestamp
            elif "timestamps" in payload and type(payload["timestamps"]) != str:
                return {"inputerror": "timestamps should be of string datatype"}
            elif "timestamps" in payload and payload ["timestamps"] != "true":
                return {"inputerror": "timestamps payload should be of Value True"}
            else:
                timestamp = True


            lang = payload.get("language")
            print(lang)
            afs = base64.b64decode(afs)

            dno = torch.cuda.current_device() if self.device == "cuda" else -1
            with tempfile.NamedTemporaryFile() as a_file:
                a_file.write(afs)
                pipe = pipeline(
                    task="automatic-speech-recognition",
                    model=self.model,
                    tokenizer=self.tokenizer,
                    feature_extractor=self.feature_extractor,
                    framework="pt",
                    chunk_length_s=30,
                    generate_kwargs={"max_new_tokens": 1024},
                    device=dno,
                    return_timestamps=timestamp
                )
                if lang:
                    self.model.config.forced_decoder_ids = self.processor.get_decoder_prompt_ids(
                        task="transcribe", language=lang
                    )
                if timestamp:
                    results = pipe(a_file.name)
                    timestamp_info = [{"text": x["text"], "start": x["timestamp"][0],"end": x["timestamp"][1]} for x in results["chunks"]]
                    return {"text": results["text"], "timestamps": timestamp_info}
                else:
                    return pipe(a_file.name)
                    
 
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "audio_file",
        type=str,
        help="Input the path to audio file you want to transcribe",
    )
    parser.add_argument(
        "--language",
        type=str,
        help="Input the language",
    )
    args = parser.parse_args()
    kwargs = vars(args)
    b64 = convert_to_base64(kwargs["audio_file"])
    if kwargs["language"]:
        payload = {"wav_base64": b64, "language": kwargs["language"]}
    payload = {"wav_base64": b64, "timestamps": "true"}
    model = Whisper()
    print(model.predict_raw(payload))
    payload = {"wav_base64": b64}
    print(model.predict_raw(payload))
```

2. When I am using model weights of `base` and `large-v2`. The code is getting an error as below:

```
Traceback (most recent call last):
  File "/root/whisper_project/microservice-ai-whisper/deploy/debug.py", line 102, in <module>
    print(model.predict_raw(payload))
  File "/root/whisper_project/microservice-ai-whisper/deploy/debug.py", line 75, in predict_raw
    results = pipe(a_file.name)
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 272, in __call__
    return super().__call__(inputs, **kwargs)
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/pipelines/base.py", line 1101, in __call__
    return next(
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/pipelines/pt_utils.py", line 124, in __next__
    item = next(self.iterator)
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/pipelines/pt_utils.py", line 266, in __next__
    processed = self.infer(next(self.iterator), **self.params)
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/pipelines/base.py", line 1015, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/pipelines/automatic_speech_recognition.py", line 445, in _forward
    tokens = self.model.generate(
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/models/whisper/modeling_whisper.py", line 1543, in generate
    return super().generate(
  File "/root/mambaforge/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/generation/utils.py", line 1406, in generate
    return self.greedy_search(
  File "/root/mambaforge/lib/python3.10/site-packages/transformers/generation/utils.py", line 2211, in greedy_search
    next_token_logits = outputs.logits[:, -1, :]
IndexError: index -1 is out of bounds for dimension 1 with size 0
```

3. I noticed this issue only with some model weights. The exact same code works with medium and large-v1 model weights

### Expected behavior

Get the transcribed output along with corresponding timestamps

### LINKED ISSUE #21937: Whisper does not respect `config.forced_decoder_ids`
### System Info

- `transformers` version: 4.27.0.dev0
- Platform: macOS-13.1-arm64-arm-64bit
- Python version: 3.9.13
- Huggingface_hub version: 0.11.1
- PyTorch version (GPU?): 1.13.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): 0.6.3 (cpu)
- Jax version: 0.4.1
- JaxLib version: 0.4.1

### Who can help?

@sanchit-gandhi @ArthurZucker

### Reproduction
Previously, the recommended method for setting the language and transcription for Whisper inference was by setting the `config.forced_decoder_ids`. However, this method no longer works on the main branch, for example if we force the model to generate in French:

```python
from transformers import WhisperProcessor, WhisperForConditionalGeneration
from datasets import load_dataset

# load model and processor
processor = WhisperProcessor.from_pretrained("openai/whisper-tiny")
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny")

ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
sample = ds[0]["audio"]
input_features = processor(sample["array"], sampling_rate=sample["sampling_rate"], return_tensors="pt").input_features 

# set the forced ids
model.config.forced_decoder_ids = processor.get_decoder_prompt_ids(language="french", task="transcribe")

# generate token ids
predicted_ids = model.generate(input_features)
# decode token ids to text
transcription = processor.batch_decode(predicted_ids, skip_special_tokens=False)
print(transcription)
```
**Print Output**:
```
['<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Mr. Quilter is the apostle of the middle classes and we are glad to welcome his gospel.<|endoftext|>']
```

### Expected behavior

Language token should correspond to the forced decoder token id (e.g. `"<|fr|>"` in this example). This was the case in previous transformer versions, so there has been a breaking change that we need to remedy.

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Should fix the backward compatibility issue with `model.config.forced_decoder_ids = ...` and should help users who want to generate with timestamps. 

Fixes #21937 and #21878

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
LGTM, thanks!

File: src/transformers/models/whisper/modeling_whisper.py [hunk 1]
```diff
@@ -1519,18 +1525,37 @@ def generate(
...
+                        f"The `{generation_config.task}`task is not supported. The task should be one of `{TASK_IDS}`"
+                    )
             else:
                 forced_decoder_ids.append((2, generation_config.task_to_id["transcribe"]))
 
-        if (
-            hasattr(generation_config, "return_timestamps") and generation_config.return_timestamps
-        ) or return_timestamps:
+        # Legacy code for backward compatibility
+        if hasattr(self.config, "forced_decoder_ids") and forced_decoder_ids != self.generation.forced_decoder_ids:
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:1549]: Should this be `generation_config.forced_decoder_ids`? We don't have an attribute `self.generation` AFAIK
```suggestion
        if hasattr(self.config, "forced_decoder_ids") and forced_decoder_ids != generation_config.forced_decoder_ids:
```

I think we should favour `forced_decoder_ids` created using the `generation_config` over the `config`. Currently, as things stand, if a user passes `language=..., task=...`, these arguments will be ignored, and the forced decoder ids will always be set based on `config.forced_decoder_ids`.

What I think we should do is only set `forced_decoder_ids = config.forced_decoder_ids` **if** neither `language` or `task` are passed.

This way, a user can still control the forced decoder ids using the generate args even if they've set `config.forced_decoder_ids`. This means older Whisper models will still be compatible with the new generate method.

WDYT?

**Review by sanchit-gandhi — COMMENTED** [review]
Amazing, thanks for fixing so promptly @ArthurZucker! My only suggestion is that we should favour `language=..., task=...` over `config.forced_decoder_ids` - see comment below!

File: src/transformers/models/whisper/modeling_whisper.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/whisper/modeling_whisper.py:1549]: Agreed! Thanks for the suggestion

File: src/transformers/models/whisper/modeling_whisper.py [hunk 2]
```diff
@@ -1504,39 +1504,60 @@ def generate(
...
-        if hasattr(generation_config, "is_multilingual") and generation_config.is_multilingual:
+        if task is not None or language is not None:
             if hasattr(generation_config, "language"):
-                forced_decoder_ids.append((1, generation_config.lang_to_id[generation_config.language]))
+                if generation_config.language in generation_config.lang_to_id.keys():
+                    language_token = generation_config.language
+                elif generation_config.language in TO_LANGUAGE_CODE.keys():
+                    language_token = f"<|{TO_LANGUAGE_CODE[generation_config.language]}|>"
+                else:
+                    raise ValueError(f"The `{generation_config.language}` language token is not supported.")
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:1531]: I get a lot of questions from people asking what format the language has to be in - should we update the docstring with the new language options and also give people the list of possible languages if they pass an invalid language? Along the lines of: https://github.com/huggingface/transformers/blob/a9bd5df16a46356463f2712dd8f6c109fa83d6f9/src/transformers/models/whisper/tokenization_whisper.py#L407-L408

File: src/transformers/models/whisper/modeling_whisper.py [hunk 3]
```diff
@@ -1504,39 +1504,60 @@ def generate(
...
-                forced_decoder_ids.append((1, None))
+                forced_decoder_ids.append((1, None))  # automatically detect the language
 
             if hasattr(generation_config, "task"):
-                forced_decoder_ids.append((2, generation_config.task_to_id[generation_config.task]))
+                if generation_config.task in TASK_IDS:
+                    forced_decoder_ids.append((2, generation_config.task_to_id[generation_config.task]))
+                else:
+                    raise ValueError(
+                        f"The `{generation_config.task}`task is not supported. The task should be one of `{TASK_IDS}`"
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:1541]: Nice

File: src/transformers/models/whisper/modeling_whisper.py [hunk 4]
```diff
@@ -1504,39 +1504,60 @@ def generate(
...
             else:
-                forced_decoder_ids.append((2, generation_config.task_to_id["transcribe"]))
-
-        if (
-            hasattr(generation_config, "return_timestamps") and generation_config.return_timestamps
-        ) or return_timestamps:
-            logits_processor = [WhisperTimeStampLogitsProcessor(generation_config)]
-        else:
-            if forced_decoder_ids and forced_decoder_ids[-1][0] != generation_config.no_timestamps_token_id:
+                forced_decoder_ids.append((2, generation_config.task_to_id["transcribe"]))  # defaults to transcribe
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:1544]: Does the official implementation also default to `transcribe` or leave the task token un-forced? (in my experience Whisper is biased to translate over transcribe if we don't force the task token, but we can follow whatever the official OpenAI implementation does here)

**Review by sanchit-gandhi — APPROVED** [review]
Awesome, thanks for working on this @ArthurZucker! LGTM!

File: src/transformers/models/whisper/modeling_whisper.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/whisper/modeling_whisper.py:1531]: yep that's exact

File: src/transformers/models/whisper/modeling_whisper.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/models/whisper/modeling_whisper.py:1544]: I think they default (from what we did with the multilingual we were already defaulting with the config)

**Comment by ArthurZucker** [comment]
The same should now be applied to both the `TF` and the `Flax` version as the overwriting of the `generate` function is also supported. Will open a follow up PR for these

**Comment by ArthurZucker** [comment]
For TF and flax #21334

**Pull request merged**